### PR TITLE
Add testrunner design documentation

### PR DESCRIPTION
**Pull request by pablochacin**
## Why is this PR needed?

As more developers contribute to the code base of the test runner, it results convenient to document its design to facilitate the understanding of the code structure, as well as to promote consistent design decisions. 

## What does this PR do?

Documents the structure of the testrunner at a high level.

## Anything else a reviewer needs to know?

This is a documentation for developers.

## Info for QA

This is info for QA so that they can validate this. This is **mandatory** if this PR fixes a bug.
If this is a new feature, a good description in "What does this PR do" may be enough.

# Merge restrictions

(Please do not edit this)

We are in *v4-maintenance phase*, so we will restrict what can be merged to prevent unexpected surprises:

    What can be merged (merge criteria):
        2 approvals:
            1 developer: code is fine
            1 QA: QA is fine
        there is a PR for updating documentation (or a statement that this is not needed)

### CODE REVIEW

File: ci/infra/testrunner/README.md [hunk 1]
```diff
@@ -30,6 +31,56 @@
 
 Testrunner is a CLI tool for setting up an environment for running e2e tests, making transparent the mechanism used for providing the test infrastructure. It can be used as a stand-alone tool, running either locally or as part of a CI pipeline. It provides commands for provisioning the infrastructure, deploying a k8s cluster using `skuba`,  and running tests. It also provides [a library](tests/README.md) for developing `pytest`-based tests.
 
+
+## Design
+
+The `testrunner` is composed by multiple components as described in the figure below:
+* The `testrunner.py` frontend cli application, which receives CLI options, loads a configuration and executes the selected command. The configuration is loaded from a yaml file (by default, [vars.yaml](vars.yaml)) and initialized by the `BaseConfig` class (located in [utils/constanst.py](utils/constanst.py) which mixes it with the environment variables and sets defaults values for missing configuration parameters.
```
Comment (ereslibre) [inline ci/infra/testrunner/README.md]: Do you think it provides value to go that detailed about the `BaseConfig` class?

File: ci/infra/testrunner/README.md [hunk 2]
```diff
@@ -30,6 +31,56 @@
...
+* A set of supporting classes which offer functionality for setting up and interacting with a cluster and executing tests. These classes wrap external tools such as `skuba`, `terraform`, `kubectl` and `pytests`. The commands offered by the `testrunner.py` cli application mostly expose the functionalities of these classes and add only the required glue code.
+* In the case of the test command, the `testrunner` not only wrap the `pytest` testing tool ([tests/driver.py](tests/driver.py), but also offers a test library ([tests/conftest.py](tests/conftest.py)) which implements reusable test functionalities or `fixtures` to facilitate [test development](tests/README.md). These fixtures use the libraries offered by testrunner for setting up the test infrastructure, deploying the cluster, interact with the cluster and the nodes, cleaning up after execution, execute common validation checks, among others.
```
Comment (ereslibre) [inline ci/infra/testrunner/README.md]:
```suggestion
* In the case of the test command, the `testrunner` does not only wrap the `pytest` testing tool ([tests/driver.py](tests/driver.py), but also offers a test library ([tests/conftest.py](tests/conftest.py)) which implements reusable test functionalities as well as `fixtures` to facilitate [test development](tests/README.md). These fixtures use the libraries offered by testrunner for setting up the test infrastructure, deploying the cluster, interact with the cluster and the nodes, cleaning up after execution, execute common validation checks, among others.
```

File: ci/infra/testrunner/README.md [hunk 3]
```diff
@@ -30,6 +31,56 @@
...
+
+The objective of this design is to maintain a clear separation between the following concerns:
+* User interface including configuration (`testrunner.py` and `BaseConfig` class)
+* Wrapping external classes (supporting libraries for `skuba`, `terraform`, `kubectl`)
+* Reusable test functionalities (`conftest.py`)
+* Test logic (diverse test suites)
+
+It worth noticing that tests can be executed directly using pytest, but it results convenient to execute them by means of the testrunner a consistent user experience.
```
Comment (ereslibre) [inline ci/infra/testrunner/README.md]:
```suggestion
It is worth noticing that tests can be executed directly using pytest, but it is more convenient to execute them using testrunner, resulting in a consistent user experience. 
```

File: ci/infra/testrunner/README.md [hunk 4]
```diff
@@ -30,6 +31,56 @@
...
+* Test logic (diverse test suites)
+
+It worth noticing that tests can be executed directly using pytest, but it results convenient to execute them by means of the testrunner a consistent user experience. 
+
+```                                           
+                                              | env variables
+                                              v
+              +--------------+         +----------------+     +---------------+
+  CLI options |              |<--------|                |     | Configuration |
+  ----------->|  Testrunner  |         | Initialiation  |<----|  (vars.yaml)  |
```
Comment (ereslibre) [inline ci/infra/testrunner/README.md]:
```suggestion
  ----------->|  Testrunner  |         | Initialization  |<----|  (vars.yaml)  |
```

File: ci/infra/testrunner/README.md (hunk 1 above)
Comment (pablochacin) [reply ci/infra/testrunner/README.md]: In practice I've found most people is not aware of this class and how it works, mostly related to how it sets defaults and merge configuration. I actually would like to rename the module to `config.py` to make more evident its intention.

Comment (pablochacin) [reply ci/infra/testrunner/README.md]: I'm removing this detail, but I think it should be documented somewhere, maybe a section on how to add new configuration parameters.

File: ci/infra/testrunner/README.md [hunk 5]
```diff
@@ -30,6 +31,56 @@
 
 Testrunner is a CLI tool for setting up an environment for running e2e tests, making transparent the mechanism used for providing the test infrastructure. It can be used as a stand-alone tool, running either locally or as part of a CI pipeline. It provides commands for provisioning the infrastructure, deploying a k8s cluster using `skuba`,  and running tests. It also provides [a library](tests/README.md) for developing `pytest`-based tests.
 
+
+## Design
+
+The `testrunner` is composed by multiple components as described in the figure below:
+* The `testrunner.py` frontend cli application, which receives CLI options, loads a configuration and executes the selected command. The configuration is loaded from a yaml file (by default, [vars.yaml](vars.yaml)) and mixed with the environment variables.
```
Comment (dannysauer) [inline ci/infra/testrunner/README.md]: Inconsistent cli v/s CLI capitalization
```suggestion
* The `testrunner.py` frontend CLI application, which receives CLI options, loads a configuration and executes the selected command. The configuration is loaded from a yaml file (by default, [vars.yaml](vars.yaml)) and mixed with the environment variables.
```

File: ci/infra/testrunner/README.md [hunk 6]
```diff
@@ -30,6 +31,56 @@
...
+* A set of supporting classes which offer functionality for setting up and interacting with a cluster and executing tests. These classes wrap external tools such as `skuba`, `terraform`, `kubectl` and `pytests`. The commands offered by the `testrunner.py` cli application mostly expose the functionalities of these classes and add only the required glue code.
+* In the case of the test command, the `testrunner` does not only wrap the `pytest` testing tool ([tests/driver.py](tests/driver.py), but also offers a test library ([tests/conftest.py](tests/conftest.py)) which implements reusable test functionalities as well as `fixtures` to facilitate [test development](tests/README.md). These fixtures use the libraries offered by testrunner for setting up the test infrastructure, deploying the cluster, interact with the cluster and the nodes, cleaning up after execution, execute common validation checks, among others.
```
Comment (dannysauer) [inline ci/infra/testrunner/README.md]:
```suggestion
* In the case of the test command, the `testrunner` not only wraps the `pytest` testing tool ([tests/driver.py](tests/driver.py), but also offers a test library ([tests/conftest.py](tests/conftest.py)) which implements reusable test functions as well as `fixtures` to facilitate [test development](tests/README.md). These fixtures use the libraries offered by testrunner for setting up the test infrastructure, deploying the cluster, interacting with the cluster and the nodes, cleaning up after execution, and executing common validation checks among others.
```
Minor grammar agreement

File: ci/infra/testrunner/README.md [hunk 7]
```diff
@@ -30,6 +31,56 @@
...
+
+The objective of this design is to maintain a clear separation between the following concerns:
+* User interface including configuration (`testrunner.py` and `BaseConfig` class)
+* Wrapping external classes (supporting libraries for `skuba`, `terraform`, `kubectl`)
+* Reusable test functionalities (`conftest.py`)
+* Test logic (diverse test suites)
+
+It is worth noticing that tests can be executed directly using pytest, but it is more convenient to execute them using the testrunner, resulting in a consistent user experience.
```
Comment (dannysauer) [inline ci/infra/testrunner/README.md]:
```suggestion
It is worth noting that, although tests can be executed directly using pytest, executing them using the testrunner is more convenient and results in a consistent user experience.
```
Minor rewording to reduce commas

File: ci/infra/testrunner/README.md (hunk 6 above)
Comment (pablochacin) [reply ci/infra/testrunner/README.md]: I had like this and @ereslibre suggested how it is now  :sweat_smile:

**Comment by pablochacin** [comment]
@dannysauer If don't mind, I would like to get the PR merged as it is, and make any improvement in further iterations. Otherwise I'll lost your approval if I push changes.

File: ci/infra/testrunner/README.md (hunk 6 above)
Comment (ereslibre) [reply ci/infra/testrunner/README.md]: Sorry, ugh.

**Comment by ereslibre** [comment]
A follow up PR can be created to address @dannysauer's feedback.

**Comment by dannysauer** [comment]
I'm pretty sure I've seen the behavior that accepting the changes inline (by clicking "apply suggestions") rather than altering your branch does not affect approvals that happened after the suggestions. I think this is because those changes get made to the PR branch itself instead of your branch.

But I can't find documentation to back that up. If you want to try it and see, I'll certainly reapprove if needed. :) Or another PR is also fine. It's up to you.

**Pull request merged**